### PR TITLE
[feature] Add extra mesh layer legend node for currently displayed vector datasets.

### LINKED ISSUE #55745: mesh layer legend item for currently displayed contour (scalar) and vector dataset
### Feature description

Vector dataset legends are not shown for mesh layers. Neither on the layer tree, nor in layout prints.
This is in spite of the fact that vector rendering can be independently set.
A mesh layer tree item must be capable of displaying both of the currently active dataset types if needed.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nstrahl**
[feature] Fixes #55745 feature request.
Mesh layer legend on the layer tree are now shown as two collapsible sub-items. One for contours, another one for vector display.

legend1

For single-color Vector rendering configuration, the arrow icon is simply re-colored in the legend (e.g. green)

legend2

This now allows for the items to be shown (and modified) in a print layout.

layout2

layout1

right-hand-side symbol display was also checked and is working:

layout3

### CODE REVIEW

**Comment by uclaros** [comment]
@nstrahl Thanks for the PR! Can you make sure the code formatting is according to the [QGIS coding standards](https://docs.qgis.org/3.28/en/docs/developers_guide/codingstandards.html)?

There's a script named `scripts/prepare_commit.sh` which is meant to be used as a pre-commit hook and should take care of properly formatting your code, so make sure you use this before committing.
For the current PR you can use `scripts/astyle_all.sh` or manually use `scripts/astyle.sh` for the modified files.

**Comment by nstrahl** [comment]
Thanks for the reply. I'm on it. I switched over to linux now, but out of curiosity how can the code be formated on Windows? I only see .sh scripts for this.

**Comment by uclaros** [comment]
Hmm, not sure about that, sorry, but I'd guess either using cygwin on wsl

File: src/core/layertree/qgslayertreemodel.cpp [hunk 1]
```diff
@@ -1471,8 +1471,33 @@ void QgsLayerTreeModel::addLegendToLayer( QgsLayerTreeLayer *nodeL )
...
+    for ( QgsLayerTreeModelLegendNode *n : nodes )
+    {
+      const QString parentName = n->property( "parentNode" ).toString();
+      QgsLayerTreeModelLegendNode *parent = rule2node.value( parentName, nullptr );
+      tree->parents[n] = parent;
+      tree->children[parent] << n;
+    }
+    return tree;
+  }
+
```
Comment (uclaros) [inline src/core/layertree/qgslayertreemodel.cpp:1500]: Duplicating this code for mesh layers and relying on custom properties feels a little hacky...
I would suggest to leave this class as is and re-use existing infrastructure for tree-like legend items.
I think you should populate `QgsLayerTreeModelLegendNode::RuleKeyRole` for the `QgsSimpleLegendNode`s and `QgsLayerTreeModelLegendNode::ParentRuleKeyRole` for the `QgsRasterSymbolLegendNode`s and `QgsColorRampLegendNode`s (you'll have to update their `data`/`setData` methods as they do not support it now) and the tree should populate itself!

File: src/core/layertree/qgslayertreemodellegendnode.cpp [hunk 2]
```diff
@@ -143,25 +143,24 @@ QSizeF QgsLayerTreeModelLegendNode::drawSymbol( const QgsLegendSettings &setting
 
   if ( ctx && ctx->painter )
   {
+    const QgsScopedRenderContextScaleToPixels scopedScaleToPixels( *( ctx->context ) );
+    const double scaleFactor = ctx->context->scaleFactor();
```
Comment (uclaros) [inline src/core/layertree/qgslayertreemodellegendnode.cpp:147]: Why do we need to multiply by the scale factor here? Won't this affect other symbols too?

File: src/core/layertree/qgslayertreemodel.cpp (hunk 1 above)
Comment (nstrahl) [reply src/core/layertree/qgslayertreemodel.cpp:1500]: Thanks for the advice. I'll  be pushing my commit in a bit. Btw, the KeyRoles are encapsulated in the classes and only settable upon construction and retrievable via the data method, so I stuck to those restrictions.

File: src/core/layertree/qgslayertreemodellegendnode.cpp (hunk 2 above)
Comment (nstrahl) [reply src/core/layertree/qgslayertreemodellegendnode.cpp:147]: Up to this point the nodes of this type have only been assigned a simple QIcon (square solid color rectangle). Under those circumstances it does not really matter if a scale factor is applied. (A square solid color rectangle more or less remains a square solid color rectangle regardless of scaling). But if a more complex QIcon is assigned (e.g. arrow) without scaling then there will be unwanted distortion/pixelation. If you look at the new /expected_legend_diagram_attributes/expected_legend_diagram_attributes_mask.png I uploaded, you will find that the icons are now slightly more rectangular because proper scaling is applied. I am guessing no one else ever bothered to do so because it was never really noticeable.

File: src/core/layertree/qgscolorramplegendnode.h [hunk 3]
```diff
@@ -43,9 +43,12 @@ class CORE_EXPORT QgsColorRampLegendNode : public QgsLayerTreeModelLegendNode
      * \param minimumLabel label text to render for the minimum value in the ramp
      * \param maximumLabel label text to render for the maximum value in the ramp
      * \param parent attach a parent QObject to the legend node.
+     * \param key rule key
+     * \param parentKey parent rule key
```
Comment (nyalldawson) [inline src/core/layertree/qgscolorramplegendnode.h:47]:
```suggestion
     * \param key rule key (since QGIS 3.36)
     * \param parentKey parent rule key (since QGIS 3.36)
```

File: src/core/layertree/qgscolorramplegendnode.h [hunk 4]
```diff
@@ -55,10 +58,13 @@ class CORE_EXPORT QgsColorRampLegendNode : public QgsLayerTreeModelLegendNode
      * \param minimumValue value associated with minimum of ramp
      * \param maximumValue value associated with maximum of ramp
      * \param parent attach a parent QObject to the legend node.
+     * \param key rule key
+     * \param parentKey parent rule key
```
Comment (nyalldawson) [inline src/core/layertree/qgscolorramplegendnode.h:62]:
```suggestion
     * \param key rule key  (since QGIS 3.36)
     * \param parentKey parent rule key (since QGIS 3.36)
```

File: src/core/layertree/qgscolorramplegendnode.h (hunk 3 above)
Comment (nyalldawson) [reply src/core/layertree/qgscolorramplegendnode.h:47]: Also would be better to add some explanatory docs here, eg for when parent rule key should be set.

File: src/core/layertree/qgslayertreemodellegendnode.h [hunk 5]
```diff
@@ -677,8 +677,10 @@ class CORE_EXPORT QgsRasterSymbolLegendNode : public QgsLayerTreeModelLegendNode
      * \param parent attach a parent QObject to the legend node.
      * \param isCheckable set to TRUE to enable the checkbox for the node (since QGIS 3.18)
      * \param ruleKey optional identifier to allow a unique ID to be assigned to the node by a renderer (since QGIS 3.18)
+     * \param parentRuleKey rule key of parent
```
Comment (nyalldawson) [inline src/core/layertree/qgslayertreemodellegendnode.h:680]:
```suggestion
     * \param parentRuleKey rule key of parent  (since QGIS 3.36)
```

Also add explanation of when this should be set

File: src/core/qgsapplication.cpp [hunk 6]
```diff
@@ -785,6 +785,11 @@ QIcon QgsApplication::getThemeIcon( const QString &name, const QColor &fillColor
     const QByteArray svgContent = QgsApplication::svgCache()->svgContent( path, 16, fillColor, strokeColor, 1, 1 );
 
     const QString iconPath = sIconCacheDir()->filePath( cacheKey + QStringLiteral( ".svg" ) );
+    if ( QDir dir = QFileInfo( iconPath ).dir(); !dir.exists() )
+    {
+      dir.mkpath( "." );
+    }
```
Comment (nyalldawson) [inline src/core/qgsapplication.cpp:791]: Can you pull this out to a separate pull request? Its unrelated to the legend changes.

File: src/core/layertree/qgslayertreemodellegendnode.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/layertree/qgslayertreemodellegendnode.cpp:147]: Watch out, `ctx->context` may be nullptr here. You'll need to check that and handle gracefully.

**Comment by nyalldawson** [comment]
Can you also add some tests covering these changes?

File: src/core/qgsapplication.cpp (hunk 6 above)
Comment (nstrahl) [reply src/core/qgsapplication.cpp:791]: Pull request #56171 has now been submitted. It will have to be accepted first so I can then rebase this branch and continue working on this pull request.

**Comment by nstrahl** [comment]
My test fails on Qt6 when comparing the expected vs. generated .png image, but it succeeds on Qt5. Can someone give me some tips? I saw that I don't have an image mask for the expected .png files so I was thinking maybe I should also add a mask. Would this solve my problem?

**Comment by nstrahl** [comment]
@uclaros Is there anything missing on this PR?

File: src/core/qgsmaplayerlegend.cpp [hunk 7]
```diff
@@ -586,13 +590,76 @@ QList<QgsLayerTreeModelLegendNode *> QgsDefaultMeshLayerLegend::createLayerTreeM
...
+                                                     shader.legendSettings() ? *shader.legendSettings() : QgsColorRampLegendNodeSettings(),
+                                                     shader.minimumValue(),
+                                                     shader.maximumValue(),
+                                                     nullptr,
+                                                     QStringLiteral( "vectorLegend" ),
+                                                     vectorNameKey );
+              }
+              break;
+            }
+            [[fallthrough]];
```
Comment (uclaros) [inline src/core/qgsmaplayerlegend.cpp:641]: Can you avoid the fallthrough? If something is to be run in all cases then I'd rather have it outside the switch statement.

**Review by uclaros — COMMENTED** [review]
The difference in the test results is quite significant. Is that caused by the introduction of the `scaleFactor` ?
Maybe it makes more sense to update the control images instead of the mask?

File: src/core/qgsmaplayerlegend.cpp (hunk 7 above)
Comment (nstrahl) [reply src/core/qgsmaplayerlegend.cpp:641]: The fallthroughs are not mine. Multiple fallthroughs throughout the project were modified/added by @nyalldawson on PR #55018 .  If the fallthroughs are an issue, then there should be a concerted effort to remove them in a separate PR. But I am of the opinion that this does not fall under the purview of this PR.

**Comment by nstrahl** [comment]
> The difference in the test results is quite significant. Is that caused by the introduction of the `scaleFactor` ? Maybe it makes more sense to update the control images instead of the mask?

The differences in `tests/testdata/control_images/legend/expected_legend_diagram_attributes/expected_legend_diagram_attributes_mask.png` are minimal. This is caused by the `scaleFactor`, but I would say the difference is negligible and even an improvement. (The layer item colored squares now look more "square")

For the brand new tests added in tests`/testdata/control_images/legend/expected_legend_mesh_diagram_no_vector`, yes the differences are a bit larger, but they arise when switching from Qt5 to Qt6. I am unsure whether the `scaleFactor` has anything to do with this. There are other tests for the mapLayerLegends and the modifications in this PR did not cause any changes in the test images. I have also opened my Qt5 and Qt6 builds and am unable to visually perceive any differences between the mesh layer legends and layouts of the two. Updating the control images as you suggested won't work since there are some small differences between Qt builds so there has to be a mask file to account for this.

**Comment by uclaros** [comment]
> Updating the control images as you suggested won't work since there are some small differences between Qt builds so there has to be a mask file to account for this.

My point here is:
- if the existing control images _no longer match the test results_, then replace them instead of updating the mask.
- If the test results differences appear between QT5 and QT6 builds, then instead of updating the mask I'd suggest to create separate control images for each:  You can create a `qt5` and a `qt6` folder within `/testdata/control_images/legend/expected_legend_mesh_diagram_no_vector` and place the respective file in each (see `/testdata/control_images/3d/expected_billboard_rendering_1/` for such an example)

File: src/core/qgsmaplayerlegend.cpp (hunk 7 above)
Comment (uclaros) [reply src/core/qgsmaplayerlegend.cpp:641]: I was referring to the actual logic that allows to fall through from one switch case to the next, however I now saw that this is copied from the `scalar` code above, so better keep it as is indeed.

File: tests/src/core/testqgslegendrenderer.cpp [hunk 8]
```diff
@@ -1435,6 +1437,78 @@ void TestQgsLegendRenderer::testDiagramAttributeLegend()
...
+  rendererSettings.setActiveScalarDatasetGroup( scalarIndex );
+  rendererSettings.setActiveVectorDatasetGroup( vectorIndex );
+  layer->setRendererSettings( rendererSettings );
+
+  std::unique_ptr< QgsLayerTree > root( std::make_unique<QgsLayerTree>() );
+  root->addLayer( layer );
+  std::unique_ptr<QgsLayerTreeModel> legendModel( std::make_unique<QgsLayerTreeModel>( root.get() ) );
+
+  QgsLegendSettings settings;
+
```
Comment (nyalldawson) [inline tests/src/core/testqgslegendrenderer.cpp:1467]: You need to call _setStandardTestFont here to normalise the legend's font, or the test won't work reliably across different platforms

Comment (nstrahl) [reply tests/src/core/testqgslegendrenderer.cpp:1467]: That was the solution that fixed the issue! 👍

**Review by uclaros — APPROVED** [review]
@nyalldawson are you OK to have this merged?

**Comment by uclaros** [comment]
@nstrahl I'm sorry for having you do this again, but can you change the `\since` mentions to `3.40`?

**Comment by uclaros** [comment]
@nstrahl can you please update the mask to resolve the conflict?

**Comment by nstrahl** [comment]
@uclaros @nyalldawson this PR is now probably ready to be merged in.

File: src/core/layertree/qgslayertreemodellegendnode.cpp [hunk 9]
```diff
@@ -142,27 +142,26 @@ QSizeF QgsLayerTreeModelLegendNode::drawSymbol( const QgsLegendSettings &setting
       size.setHeight( ctx->patchSize.height( ) );
   }
 
-  if ( ctx && ctx->painter )
+  if ( ctx && ctx->painter && ctx->context )
   {
+    const QgsScopedRenderContextScaleToPixels scopedScaleToPixels( *( ctx->context ) );
+    const double scaleFactor = ctx->context ? ctx->context->scaleFactor() : 1;
```
Comment (nyalldawson) [inline src/core/layertree/qgslayertreemodellegendnode.cpp:148]:
```suggestion
    const double scaleFactor = ctx->context->scaleFactor();
```

**Review by nyalldawson — APPROVED** [review]
